Build watch

frandiox opened the pull request:
This is a prerequisite for supporting `h2 debug cpu` in Vite. Features:

- Add `--watch` to `h2 build`
- Add `--build --watch --codegen` to `h2 preview`
- Support `--diff` in `h2 preview`

## Note

Miniflare is currently not compatible with CSP nonces and WS connections. Therefore, live reload doesn't work with `h2 preview --build --watch` until that's fixed and you need to refresh the browser manually.
I've reported the problem here: https://github.com/cloudflare/workers-sdk/issues/5829

frandiox commented on packages/cli/src/commands/hydrogen/build.ts:
```diff
@@ -197,4 +258,16 @@ export async function runBuild({
+      codegenProcess?.kill('SIGINT');
+
+      const promises: Array<Promise<void>> = [];
+      if ('close' in clientBuild) promises.push(clientBuild.close());
+      if ('close' in serverBuild) promises.push(serverBuild.close());
```
The type of `clientBuild` and `serverBuild` is different in watch mode. Using `in` here correctly discriminates the type.

frandiox commented on packages/cli/src/commands/hydrogen/preview.ts:
```diff
@@ -31,47 +40,135 @@ export default class Preview extends Command {
     });
   }
 }
 
 type PreviewOptions = {
   port?: number;
-  path?: string;
+  directory?: string;
   legacyRuntime?: boolean;
   env?: string;
   envBranch?: string;
   inspectorPort?: number;
   debug: boolean;
   verbose?: boolean;
+  build?: boolean;
+  watch?: boolean;
+  entry?: string;
+  codegen?: boolean;
+  codegenConfigPath?: string;
 };
 
 export async function runPreview({
   port: appPort,
-  path: appPath,
+  directory,
   legacyRuntime = false,
   env: envHandle,
   envBranch,
   inspectorPort,
   debug,
   verbose,
+  build: shouldBuild = false,
+  watch = false,
+  codegen: useCodegen = false,
+  codegenConfigPath,
+  entry,
 }: PreviewOptions) {
-  if (!process.env.NODE_ENV) process.env.NODE_ENV = 'production';
+  if (!process.env.NODE_ENV)
+    process.env.NODE_ENV = watch ? 'development' : 'production';
```
Using `development` with watch mode to allow our CSP to connect to Miniflare's loopback server (localhost:<random>).

blittle commented:
I can only repro this issue 50% of the time or so, but when running  `h2 build --watch --diff` inside `infinite-scroll` example. I modify a file, it rebuilds, and then I try and kill the process (ctr + c) but the process doesn't die. It just infinitely keeps printing this error:

![image](https://github.com/Shopify/hydrogen/assets/1566869/4427e62d-38fa-4bfd-8e8f-f0effe40ff1a)

frandiox commented:
>I can only repro this issue 50% of the time or so, but when running h2 build --watch --diff inside infinite-scroll example. I modify a file, it rebuilds, and then I try and kill the process (ctr + c) but the process doesn't die. It just infinitely keeps printing this error:

@blittle  I was only able to reproduce this very few times. It looks like even when we close Rollup watchers, the build still continues and then it starts failing synchronously when it can't find the diff files that we just removed on process exit.

I've added a few changes to try to ensure the build exits before we remove files so that we don't go into an infinite sync error loop. I'm not able to reproduce it after these changes but maybe I was just lucky. Please try it again but, in any case, these errors only happens with `--diff` so it's probably not a big issue.

blittle reviewed (approved):
I still see the `--diff` problem, but like you said, it only affects us.
